In [7]:
import pandas as pd

### Process the Genres

In [1]:
# Read books table
df = pd.read_csv("C:/Users/Christina/OneDrive - Hochschule Luzern/DBM Project/Data/books_1.Best_Books_Ever.csv")

NameError: name 'pd' is not defined

In [9]:
# Delete columns not needed for the genre table
genre = df.drop(["bookId", "title", "series", "author", "rating", "description", "language", "characters", "bookFormat", "edition", "pages", "publisher", "publishDate", "firstPublishDate", "awards", "numRatings", "ratingsByStars", "likedPercent", "setting", "coverImg", "bbeScore", "bbeVotes", "price"], axis=1)

In [10]:
# Drop all books without valid isbn
genre.drop(genre[genre['isbn'] == "9999999999999"].index, inplace = True)

In [11]:
# Split the rows with repeating groups into separate rows
genre = genre.assign(genres=genre.genres.str.split(',')).explode('genres')

In [12]:
# Save new table into csv
genre.to_csv("C:/Users/Christina/OneDrive - Hochschule Luzern/DBM Project/Data/genres.csv", index=False)

Note: further cleaning took place in Excel

In [32]:
# Upload cleaned genre table
genre2 = pd.read_csv("C:/Users/Christina/OneDrive - Hochschule Luzern/DBM Project/Data/genres_new.csv", sep=";")

In [33]:
# Get amount of values
genre2

isbn           genre
0       9780439023481      YoungAdult
1       9780439023481         Fiction
2       9780439023481        Dystopia
3       9780439023481         Fantasy
4       9780439023481  ScienceFiction
...               ...             ...
382181  9781599554976             Lds
382182  9781599554976      Historical
382183  9781599554976         Romance
382184  9781599554976       Adventure
382185  9781599554976            Teen

[382186 rows x 2 columns]

In [34]:
# Amount of unique values
genre2['genre'].nunique()

981

There are 382186 rows with 981 unique values in the genre table

In [37]:
# Separate Genres into new table
genre_id = pd.DataFrame(genre2['genre'].unique(), columns=["genre"])

In [39]:
# Add unique ID to genres
genre_id.insert(0, 'Genre_ID', range(1, 1 + len(genre_id)))
genre_id

Genre_ID              genre
0           1         YoungAdult
1           2            Fiction
2           3           Dystopia
3           4            Fantasy
4           5     ScienceFiction
..        ...                ...
976       977  LondonUnderground
977       978           FarRight
978       979            PeakOil
979       980           Airships
980       981             Guinea

[981 rows x 2 columns]

In [40]:
# Replace genre with genre ID in genre2
genre2['genre'] = genre2['genre'].map(genre_id.set_index('genre')['Genre_ID'])
genre2

isbn  genre
0       9780439023481      1
1       9780439023481      2
2       9780439023481      3
3       9780439023481      4
4       9780439023481      5
...               ...    ...
382181  9781599554976    263
382182  9781599554976     23
382183  9781599554976      6
382184  9781599554976      7
382185  9781599554976      8

[382186 rows x 2 columns]

In [42]:
# Rename the column
genre2.rename(columns = {'genre':'genre_id'}, inplace = True)
genre2

isbn  genre_id
0       9780439023481         1
1       9780439023481         2
2       9780439023481         3
3       9780439023481         4
4       9780439023481         5
...               ...       ...
382181  9781599554976       263
382182  9781599554976        23
382183  9781599554976         6
382184  9781599554976         7
382185  9781599554976         8

[382186 rows x 2 columns]

In [43]:
# Save both tables
genre2.to_csv("C:/Users/Christina/OneDrive - Hochschule Luzern/DBM Project/Data/book_genres_final.csv", index=False)
genre_id.to_csv("C:/Users/Christina/OneDrive - Hochschule Luzern/DBM Project/Data/genres_ids.csv", index=False)

### Repeat the process for the awards

In [14]:
# Getting rid of not needed columns for the awards table
award = df.drop(["bookId", "title", "series", "author", "rating", "description", "language", "characters", "bookFormat", "edition", "pages", "publisher", "publishDate", "firstPublishDate", "genres", "numRatings", "ratingsByStars", "likedPercent", "setting", "coverImg", "bbeScore", "bbeVotes", "price"], axis=1)

In [15]:
# Drop all books without valid isbn
award.drop(award[award['isbn'] == "9999999999999"].index, inplace = True)

In [16]:
# Split the rows with repeating groups into separate rows
award = award.assign(awards=award.awards.str.split(',')).explode('awards')

In [17]:
# Save new awards table into csv
award.to_csv("C:/Users/Christina/OneDrive - Hochschule Luzern/DBM Project/Data/awards.csv", index=False)

Note: here, too, further cleaning took place in Excel

In [29]:
# Upload cleaned awards table
award2 = pd.read_csv("C:/Users/Christina/OneDrive - Hochschule Luzern/DBM Project/Data/awards_new.csv", sep=";")

In [30]:
award2

isbn                                              award
0      9780439023481             LocusAwardNomineeforBestYoungAdultBook
1      9780439023481                              GeorgiaPeachBookAward
2      9780439023481                                    BuxtehuderBulle
3      9780439023481                       GoldenDuckAwardforYoungAdult
4      9780439023481  GrandPrixdelImaginaireNomineeforRomanjeunesseé...
...              ...                                                ...
21541  9781760111236        NitaB.KibbleLiteraryAwardNomineeforLonglist
21542  9781760111236               VossLiteraryPrizeNomineeforShortlist
21543  9781111493820             WorldFantasyAwardNomineeforBestNovella
21544  9781414348643  HOLTMedallionbyVirginiaRomanceWritersNomineefo...
21545  9781461017097                           ReadersFavoriteBookAward

[21546 rows x 2 columns]

In [31]:
# Amount of unique values
award2['award'].nunique()

5080

There are 21546 rows with 5080 unique values in the awards table

In [44]:
# Separate Awards into new table
award_id = pd.DataFrame(award2['award'].unique(), columns=["award"])

In [45]:
# Add unique ID to awards
award_id.insert(0, 'Award_ID', range(1, 1 + len(award_id)))
award_id

Award_ID                                              award
0            1             LocusAwardNomineeforBestYoungAdultBook
1            2                              GeorgiaPeachBookAward
2            3                                    BuxtehuderBulle
3            4                       GoldenDuckAwardforYoungAdult
4            5  GrandPrixdelImaginaireNomineeforRomanjeunesseé...
...        ...                                                ...
5075      5076                        ColoradoBookAwardforMystery
5076      5077                                AmberButterflyPrize
5077      5078         VermontsPictureBookAwards:RedCloverNominee
5078      5079  MilesFranklinLiteraryAwardNomineeforLiteraryFi...
5079      5080        NitaB.KibbleLiteraryAwardNomineeforLonglist

[5080 rows x 2 columns]

In [46]:
# Replace award with award ID in award2
award2['award'] = award2['award'].map(award_id.set_index('award')['Award_ID'])
award2

isbn  award
0      9780439023481      1
1      9780439023481      2
2      9780439023481      3
3      9780439023481      4
4      9780439023481      5
...              ...    ...
21541  9781760111236   5080
21542  9781760111236   1384
21543  9781111493820    622
21544  9781414348643   1977
21545  9781461017097   2120

[21546 rows x 2 columns]

In [47]:
# Rename the column
award2.rename(columns = {'award':'award_id'}, inplace = True)
award2

isbn  award_id
0      9780439023481         1
1      9780439023481         2
2      9780439023481         3
3      9780439023481         4
4      9780439023481         5
...              ...       ...
21541  9781760111236      5080
21542  9781760111236      1384
21543  9781111493820       622
21544  9781414348643      1977
21545  9781461017097      2120

[21546 rows x 2 columns]

In [48]:
# Save both tables
award2.to_csv("C:/Users/Christina/OneDrive - Hochschule Luzern/DBM Project/Data/book_awards_final.csv", index=False)
award_id.to_csv("C:/Users/Christina/OneDrive - Hochschule Luzern/DBM Project/Data/awards_ids.csv", index=False)